src/app/services/contracts.service.ts

```diff
@@ -0,0 +1,156 @@
+    decimals: number;
+    underlyingUnit: number;
+    isStableCoin() {
+        return this.underlying?.type === 4 && !this.contract?.name?.match(/CRV/);
+    }
+    isCRV() {
+        return this.underlying?.type === 4 && this.contract?.name?.match(/CRV/);
+    }
+    isUniLP() {
+        return this.underlying?.type === 2 && this.underlying?.name?.match(/^UNI_LP/);
+    }
+    isOtherLP() {
+        return this.underlying?.type === 2 && !this.underlying?.name?.match(/^UNI_LP/);
+    }
+}
+
+export class Pool implements IsContract {
+    id: number;
+    contract: Contract;
+    updatedBlock: number;
+    controller: Contract;
+    governance: Contract;
+    owner: Contract;
+    lpToken: Contract;
+    rewardToken: Contract;
+}
+
+export class Token implements IsContract {
+    id: number;
+    contract: Contract;
+    updatedBlock: number;
+    name: string;
+    symbol: string;
+    decimals: number;
+}
+
+export class Pair implements IsContract {
+    id: number;
+    contract: Contract;
+    updatedBlock: number;
+    keyToken: Token;
+    type: number;
+    decimals: number;
+    token0: Contract;
+    token1: Contract;
+}
+
+interface Service<T extends IsContract> {
+    getContract(type: new () => T, name: string): Observable<T>;
+    getContracts(type: new () => T): Observable<T[]>;
+}
+
+const TYPE_PATHS = new Map<IsContract, string>([[Vault, 'vault'], [Pool, 'pool'], [Token, 'token'], [Pair, 'unipair']]);
+
+@Injectable({
+    providedIn: 'root'
+})
+export class ContractsService<T> implements Service<T> {
+
+    private urlPrefix = 'contracts';
+
+    constructor(private http: HttpClient, private snackService: SnackService) {
+
+    }
+
+    getContract(type: new () => T, name: string): Observable<T> {
+        return this.http.get(`${environment.apiEndpoint}/${this.urlPrefix}/${TYPE_PATHS.get(type)}/${name}`).pipe(
+            catchError(this.snackService.handleError<Result<T>>(`Contract fetch for ${TYPE_PATHS.get(type)} ${name} failed.`)),
+            map((val: Result<T>) => Object.assign(new type(), val.data) as T)
+        );
+    }
+
+    getContracts(type: new () => T): Observable<T[]> {
+        return this.http.get(`${environment.apiEndpoint}/${this.urlPrefix}/${TYPE_PATHS.get(type)}s`).pipe(
+            catchError(this.snackService.handleError<Result<T>>(`Contracts fetch for ${TYPE_PATHS.get(type)} failed.`)),
+            map((val: Result<T>) => (val.data as T[]).map(o => Object.assign(new type(), o)) as T[]),
+            map(_ => _.filter(item => !(item instanceof Vault) || !(item.contract?.name?.match(/_V0$/))))
+        );
+    }
+
+}
+
+@Injectable({
+    providedIn: 'root'
+})
+export class StaticContractsService implements Service<Vault|Pool> {
+
+    private readonly vaults: Vault[];
+    private index: Map<string, number> = new Map();
+
+    constructor() {
+        this.vaults = StaticValues.currentVaults.map(v => ({contract: {name: v, address: Addresses[v] || Addresses[`_${v}`]}} as Vault));
```

Comment:
addresses.ts is deprecated and should be removed
don't use static values
why we need so complex structure StaticContractsService -> ContractService ->Service
in your plans create other implementations?
don't use an interface for single implementation without good reason